sympy/crypto/crypto.py

```diff
@@ -2037,3 +2039,155 @@ def dh_shared_key(key, b):
             than prime %s.''' % p))
 
     return pow(x, b, p)
+
+
+################ Goldwasser-Micali Encryption  #########################
+
+
+def legendre(a, p):
```

Comment:
It would be preferable to use `legendre_symbol` that is implemented in the `ntheory` module.

Reply:
The `legendre symbol` in `ntheory` does a prime check at each call. This means that we will be doing a prime check for large `p` and `q` for every bit we send. if this is something you are ok with (since the encryption module in crypto is most likely meant to be used in research and tests, not in actual production), then I will change it, but do confirm please.

Reply:
> The legendre symbol in ntheory does a prime check at each call.

That will justify having a local function. Another observation: it would probably be more efficient to use the three argument `pow(a%p, (p - 1)//2, p)` instead of `pow(a%p, (p - 1)//2) % p`.